### PR TITLE
Custom groups

### PR DESCRIPTION
**Pull request by shaver**
As discussed in the forum thread, this adds the ability to create and manage custom groups and the dwarves who love them.

Right click a dwarf name to add a new group, delete an existing one, or add/remove from a group. Works with multiple selections.

Group names are global, but dwarf/group assignments are per-fortress.

(Also has the fix for OS X and setuid detection; I can split that out if you want.)

I've never done any Qt before, so please don't be shy about pointing out where I've been dumb.

### CODE REVIEW

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -361,6 +362,7 @@ void MainWindow::connect_to_df() {
             if (DT->user_settings()->value("options/read_on_startup", true).toBool()) {
                 read_dwarves();
             }
+            DT->on_connect();
```
Comment (splintermind) [inline src/mainwindow.cpp]: Consider changing this to emit a signal from <code>read_dwarves</code> instead. Currently <code>on_connect</code> will be called even if <code>read_dwarves</code> fails.

File: inc/models/dwarfmodel.h [hunk 2]
```diff
@@ -59,6 +43,7 @@ class DwarfModel : public QStandardItemModel {
         GB_CASTE,
         GB_CASTE_TAG,
         GB_CURRENT_JOB,
+        GB_CUSTOM_GROUP,
```
Comment (splintermind) [inline inc/models/dwarfmodel.h]: Move this to the end of the enum. These are cast to integers to use as an id number when saving group by, so inserting an item will break any saved group settings.

File: src/dwarftherapist.cpp [hunk 3]
```diff
@@ -265,6 +273,80 @@ void DwarfTherapist::save_custom_prof(CustomProfession *cp){
...
+
+    const QString &fortid = get_DFInstance()->fortress_unique_identifier();
+    count = m_user_settings->beginReadArray(fortid);
+    for (i = 0; i < count; i++) {
+        m_user_settings->setArrayIndex(i);
+        const QString &name = m_user_settings->value("name").toString();
+        CustomGroup *g = get_custom_group(name);
+        if (!g) // has been deleted when connected to a different fort
+            continue;
+        const QString &memberstring = m_user_settings->value("members").toString();
```
Comment (splintermind) [inline src/dwarftherapist.cpp]: Consider passing the <code>QSettings</code> into the <code>CustomGroup</code> object to move the loading into the object itself. See <code>labor.h</code> for a simple example.

File: src/models/dwarfmodel.cpp [hunk 4]
```diff
@@ -345,6 +346,17 @@ void DwarfModel::build_rows() {
                         m_grouped_dwarves[tr("Losers")].append(d);
                 }else if(m_group_by == GB_HAPPINESS){
                     m_grouped_dwarves[d->happiness_name(d->get_happiness())].append(d);
+                } else if(m_group_by == GB_CUSTOM_GROUP){
+                    bool is_grouped = false;
+                    foreach (CustomGroup *g, DT->get_custom_groups()) {
+                        if (g->has_member(d)) {
```
Comment (splintermind) [inline src/models/dwarfmodel.cpp:356]: As an alternative to looping through every group and it's members, when you assign the dwarves to groups, assign each dwarf the particular custom group name they belong to. Then you can just call <code>d->get_custom_group_name()</code>.

File: src/statetableview.cpp [hunk 5]
```diff
@@ -512,6 +546,68 @@ void StateTableView::set_squad_name(){
...
 
+void StateTableView::add_custom_group(){
+    bool ok;
+    QString new_group = QInputDialog::getText(this, tr("New Custom Group"),
+                                              tr("Enter name for new custom group."), QLineEdit::Normal,
+                                              tr(""), &ok);
+    if (!ok)
+        return;
+
+    DT->add_custom_group(new_group);
```
Comment (splintermind) [inline src/statetableview.cpp]: Would be nice to also pass in the list of currently selected dwarves and immediately add them to the new group.

**Comment by splintermind** [comment]
I'm concerned that the global settings file may not be the best place to save this information, as over time it will bloat with groups from every fortress you've ever had.

Is there a reason you didn't go with your original concept of storing an additional file in the save directory itself?

Other than that, it might be useful to have another dock window to show the currently loaded custom groups, remove them, add new ones, and perhaps edit them (at least to change the name).

File: src/models/dwarfmodel.cpp (hunk 4 above)
Comment (shaver) [inline src/models/dwarfmodel.cpp:356]: Dwarves can be in multiple groups, but I could keep a vector there. I'll see how it feels.

File: src/dwarftherapist.cpp (hunk 3 above)
Comment (shaver) [inline src/dwarftherapist.cpp]: One challenge is that we don't want to deserialize cases where the global group has been deleted while running a different fort. This means that the deserialization logic needs to know about the global list, and I didn't want to have groups talk to DwarfTherapist. I can encapsulate it a bit better, though.

**Comment by shaver** [comment]
I'd like to do the save directory thing, yeah. I'll try to figure out the offsets, but I wanted to be able to test and get feedback before I did that.

Would you be OK deferring dock/rename for a later time? It's not too hard to select a group, add a new one, and then delete the old, for now.

**Comment by splintermind** [comment]
So where are you at with this, is it pretty much ready to go? I've probably broken everything with the recent merge clusterfuck from superlabors into DF2014 but it shouldn't be too bad even if it has to be manually merged.

**Comment by shaver** [comment]
I have it rebased against DF2014 and just need to figure out the offsets.
I'll give it a shot tonight, should have a bit of time.

**Pull request closed without merging**